pkg/integrity/dsse.go

```diff
@@ -0,0 +1,212 @@
+func newDSSEEncoder(ss []signature.Signer, opts ...signature.SignOption) (*dsseEncoder, error) {
+	var so crypto.SignerOpts
+	for _, opt := range opts {
+		opt.ApplyCryptoSignerOpts(&so)
+	}
+
+	// If SignerOpts not explicitly supplied, set default hash algorithm.
+	if so == nil {
+		so = crypto.SHA256
+		opts = append(opts, options.WithCryptoSignerOpts(so))
+	}
+
+	dss := make([]dsse.SignVerifier, 0, len(ss))
+	for _, s := range ss {
+		ds, err := newDSSESigner(s, opts...)
+		if err != nil {
+			return nil, err
+		}
+
+		dss = append(dss, ds)
+	}
+
+	es, err := dsse.NewEnvelopeSigner(dss...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &dsseEncoder{
+		es:          es,
+		h:           so.HashFunc(),
+		payloadType: metadataMediaType,
+	}, nil
+}
+
+// signMessage signs the message from r in DSSE format, and writes the result to w. On success, the
+// hash function is returned.
+func (en *dsseEncoder) signMessage(w io.Writer, r io.Reader) (crypto.Hash, error) {
+	body, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	e, err := en.es.SignPayload(en.payloadType, body)
+	if err != nil {
+		return 0, err
+	}
+
+	return en.h, json.NewEncoder(w).Encode(e)
+}
+
+type dsseDecoder struct {
+	vs          []signature.Verifier
+	threshold   int
+	payloadType string
+}
+
+// newDSSEDecoder returns a decoder that verifies messages in DSSE format using key material from
+// vs.
+func newDSSEDecoder(vs ...signature.Verifier) *dsseDecoder {
+	return &dsseDecoder{
+		vs:          vs,
+		threshold:   1, // Envelope considered verified if at least one verifier succeeds.
+		payloadType: metadataMediaType,
+	}
+}
+
+var (
+	errDSSEVerifyEnvelopeFailed  = errors.New("dsse: verify envelope failed")
+	errDSSEUnexpectedPayloadType = errors.New("unexpected DSSE payload type")
+)
+
+// verifyMessage reads a message from r, verifies its signature(s), and returns the message
+// contents. On success, the accepted public keys are set in vr.
+func (de *dsseDecoder) verifyMessage(r io.Reader, h crypto.Hash, vr *VerifyResult) ([]byte, error) {
+	vs := make([]dsse.Verifier, 0, len(de.vs))
+	for _, v := range de.vs {
+		dv, err := newDSSEVerifier(v, options.WithCryptoSignerOpts(h))
+		if err != nil {
+			return nil, err
+		}
+
+		vs = append(vs, dv)
+	}
+
+	v, err := dsse.NewMultiEnvelopeVerifier(de.threshold, vs...)
+	if err != nil {
+		return nil, err
+	}
+
+	var e dsse.Envelope
+	if err := json.NewDecoder(r).Decode(&e); err != nil {
+		return nil, err
+	}
+
+	vr.aks, err = v.Verify(&e)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", errDSSEVerifyEnvelopeFailed, err)
+	}
+
+	if e.PayloadType != de.payloadType {
+		return nil, fmt.Errorf("%w: %v", errDSSEUnexpectedPayloadType, e.PayloadType)
+	}
+
+	return e.DecodeB64Payload()
+}
+
+type dsseSigner struct {
+	s    signature.Signer
+	opts []signature.SignOption
+	pub  crypto.PublicKey
+}
+
+// newDSSESigner returns a dsse.SignVerifier that uses s to sign according to opts.
```

Comment:
Should the comment here be explicit that though it's a SignVerifier... it doesn't implement the Verifier part of the interface?